base/setup/lib/utils/volutil.c

```diff
@@ -0,0 +1,207 @@
+
+    /* Retrieve the volume label */
+    if (VolumeHandle)
+    {
+        IO_STATUS_BLOCK IoStatusBlock;
+        UCHAR LabelBuffer[sizeof(FILE_FS_VOLUME_INFORMATION) + 256 * sizeof(WCHAR)];
+        PFILE_FS_VOLUME_INFORMATION LabelInfo = (PFILE_FS_VOLUME_INFORMATION)LabelBuffer;
+
+        Status = NtQueryVolumeInformationFile(VolumeHandle,
+                                              &IoStatusBlock,
+                                              &LabelBuffer,
+                                              sizeof(LabelBuffer),
+                                              FileFsVolumeInformation);
+        if (NT_SUCCESS(Status))
+        {
+            /* Copy the (possibly truncated) volume label and NULL-terminate it */
+            RtlStringCbCopyNW(Volume->VolumeLabel, sizeof(Volume->VolumeLabel),
+                              LabelInfo->VolumeLabel, LabelInfo->VolumeLabelLength);
+        }
+        else
+        {
+            DPRINT1("NtQueryVolumeInformationFile() failed, Status 0x%08lx\n", Status);
+        }
+    }
+
+    /* Close the volume */
+    if (VolumeHandle)
+        NtClose(VolumeHandle);
+
+    return STATUS_SUCCESS;
+}
+
+/**
+ * @brief
+ * Attempts to dismount the designated volume.
+ **/
+NTSTATUS
+DismountVolume(
+    _Inout_ PVOLINFO Volume)
+{
+    NTSTATUS Status, LockStatus;
+    IO_STATUS_BLOCK IoStatusBlock;
+    HANDLE VolumeHandle;
+
+    /* Check whether the volume was mounted by the system.
+     * If the volume is not mounted, just return success. */
+    if (!*Volume->FileSystem) // (IsUnknown(Volume))
+        return STATUS_SUCCESS;
+
+    /* Open the volume */
+    Status = pOpenDeviceEx(Volume->DeviceName, &VolumeHandle,
+                           GENERIC_READ | GENERIC_WRITE,
+                           FILE_SHARE_READ | FILE_SHARE_WRITE);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("ERROR: Cannot open volume %S for dismounting! (Status 0x%lx)\n",
+                Volume->DeviceName, Status);
+        return Status;
+    }
+
+    /* Lock the volume */
+    LockStatus = NtFsControlFile(VolumeHandle,
+                                 NULL, NULL, NULL,
+                                 &IoStatusBlock,
+                                 FSCTL_LOCK_VOLUME,
+                                 NULL, 0,
+                                 NULL, 0);
+    if (!NT_SUCCESS(LockStatus))
+    {
+        DPRINT1("WARNING: Failed to lock volume, operations may fail! (Status 0x%lx)\n", LockStatus);
+    }
+
+    /* Dismount the volume */
+    Status = NtFsControlFile(VolumeHandle,
+                             NULL, NULL, NULL,
+                             &IoStatusBlock,
+                             FSCTL_DISMOUNT_VOLUME,
+                             NULL, 0,
+                             NULL, 0);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("Failed to unmount volume (Status 0x%lx)\n", Status);
+    }
+
+    /* Unlock the volume */
+    LockStatus = NtFsControlFile(VolumeHandle,
```

Comment:
This would probably be unnecessary; I remember @hpoussin made some code change in the setup code in the past where he was unlocking only if locking succeeded before.

Reply:
Addressed